lib/posix-process: Add `signalfd`/`signalfd4` support

mogasergiu opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.


### Base target

 - Architecture(s): [e.g. `x86_64` or N/A]
 - Platform(s): [e.g. `kvm`, `xen` or N/A]
 - Application(s): [e.g. `app-python3` or N/A]


### Additional configuration

<!--
Please specify any additional configuration which is needed for this feature to
work or any new configuration parameters which are introduced by this PR.  This
will help during the review process.  For example:

 - `CONFIG_LIBUKDEBUG=y`

-->

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->

Add support for the `signalfd`/`signalfd4` syscalls which will     
allow callers to add monitored signal files to the current process,
allowing them to be notified of pending signals.

mogasergiu committed:
lib/posix-process: Add `signalfd`/`signalfd4` support

Add support for the `signalfd`/`signalfd4` syscalls which will
allow callers to add monitored signal files to the current process,
allowing them to be notified of pending signals.

Introduce the primitives required for supporting signalfd: signal files.

These embed the actual base file structure, the mask that shows what
signals are being monitored as well as other file data to properly
register with libukfile.

Each process' signal descriptor holds control over the signal files
context which tracks the combined mask of all of the open signal files
as well as a list of their references. Having a combined mask helps
to make checking to see whether a process has signal files monitoring
for a given signal faster as compared to iterating every time through
each signal file and checking their masks.

Thus, adding a new signal file entry means updating the combined mask.
Deleting one implies resetting the signal mask and reiterating
through all remaining signal files for a proper combined mask. However,
as an optimization, we don't do that. Instead, we defer mask
recalculation until we need to reiterate through the signal files
again, which is when we have to notify the registered signal files
against a signal number.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu committed:
lib/posix-process: Init signal files ctx on signal desc creation

We have two initialization places for the signal files context of a
signal files context:
- on vfork when a child process is born and its signal descriptor is
correspondingly inherited from the parent
- on signal descriptor initialization on regular process creation

For the former, at this time, we do not support proper open file
inheritance across vfork/execve for those that do not have the
O_CLOEXEC flag. Therefore, for now, simply just initialize child
process' signal descriptor signal file context as empty.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu committed:
lib/posix-process: Notify process/thread signal files when possible

Whenever a signal shouldn't be dropped, meaning it is either not ignored by
the process or the process has signal files that are monitoring that
particular signal, notify the existing signal files in question
 of the posix thread/process by setting the input event on said files.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu committed:
lib/nolibc: Import `signalfd` definitions from `musl`

Import `musl`'s header for `signalfd` definitions to allow
nolibc builds a way to call into the signalfd implementation.

Source of import:
Repository: https://git.musl-libc.org/cgit/musl
Tag: v1.2.4 (f5f55d65)
Path: include/sys/signalfd.h

Checkpatch-Ignore: SPDX_LICENSE_TAG
Checkpatch-Ignore: FUNCTION_ARGUMENTS
Checkpatch-Ignore: PREFER_KERNEL_TYPES
Checkpatch-Ignore: SPACING
Checkpatch-Ignore: COMMIT_LOG_USE_LINK
Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

andreittr commented on lib/posix-process/signal/signal.h:
```diff
@@ -9,6 +9,7 @@
 
 #include <limits.h>
 #include <signal.h>
+#include <stdbool.h>
```
Do the signal file additions require this header? I don't see any `bool`s being used there :thinking: 
Or is this a fix for an unrelated bool use somewhere else in the file?

mogasergiu replied:
Nope, this should've gone in signal_file.h
The changes had a very different placement initially, everything was defined in signalfd.c while testing/developing.

mogasergiu replied:
Oh nevermind, it is indeed used in a later commit. I'll add it in both files in the separate commits.

andreittr commented on lib/posix-process/signal/signal_file.h:
```diff
@@ -0,0 +1,107 @@
+#endif
+
+#include <uk/file.h>
+#include <uk/posix-fd.h>
```
I believe you can remove the `posix-fd` include; this file doesn't look like it uses any types/funcs defined there.

mogasergiu replied:
Hmm, seems I included it because of UKFD_POLLIN - so it should go into signal.h instead

andreittr commented on lib/posix-process/signal/signal_file.h:
```diff
@@ -0,0 +1,107 @@
+struct uk_signal_file *uk_file_to_signal_file(const struct uk_file *f)
+{
+	return __containerof(f, struct uk_signal_file, f);
+}
```
As a file driver you have complete freedom over the values in the `.vol` and `.node`, and I noticed later in the implementation that you store the mask in node.
That's perfectly fine, but you could get even cleaner code by storing the `struct uk_signal_file *` in `f->node` instead; that way `f->node` provides all the internal runtime context of a sigfile for `read()`, `poll()`, and signal delivery ops.

Of course it's a design freedom, and it's up to you how you structure things. I just want to make sure you're aware of all possibilities.

mogasergiu replied:
Sounds ok to me. I even had it like this initially but then I changed it to follow the pattern of timerfd and only have in node the private data that matters. Will change :+1:.

andreittr commented on lib/posix-process/signal/signalfd.c:
```diff
@@ -0,0 +1,345 @@
+	uk_sigset_t mask;
+
+	if (unlikely(!FILE_IS_SIGNAL(f)))
+		return -EINVAL;
```
You can safely turn this into an assert.

The only way this function is called with `f` is if `f->ops->read == sigfile_read`. Because both files as well as their ops tables are immutable once created, this can only happen if a non-signal file has signal file ops -- indicative of either memory corruption or an unambiguous kernel bug that is not recoverable.

andreittr commented on lib/posix-process/signal/signalfd.c:
```diff
@@ -0,0 +1,345 @@
+
+static ssize_t sigfile_read(const struct uk_file *f,
+			    const struct iovec *iov, size_t iovcnt __unused,
+			    size_t off __unused, long flags __unused)
```
`iovcnt` and `off` are being used in args validation, you can remove the `__unused` attrs.

mogasergiu replied:
Ah, right, I copy-pasted the checks from eventfd.c after writing the function :sweat_smile:

andreittr commented on lib/posix-process/signal/signalfd.c:
```diff
@@ -0,0 +1,345 @@
+	/*
+	 * signalfd is only meant for read() base syscall variant so there
+	 * is only one buffer.
+	 */
```
While I (think I) understand what you meant here -- signalfd's `read` can only output into one contiguous buffer by design -- the wording suggests that it wouldn't work with other read variants even if given the correct args (which I _hope_ is not the case).

mogasergiu replied:
Will update comment to make it more clear. What I was referring to was that we are assuming we are not just called in a readv with split buffer.

andreittr replied:
Ok, but is this assumption documented (or at least observed) in Linux, or is it a limitation we introduce? Either way is fine, but the comment must explicitly say which.

mogasergiu replied:
Heh, this is funny, I was about to say "observed in linux"... but the version I used for reference did not have this [commit](https://lore.kernel.org/linux-kernel/8de27ff2-d53f-4900-b755-6a05bff98f78@kernel.dk/T/)... will change to use the iov API.

andreittr commented on lib/posix-process/signal/signalfd.c:
```diff
@@ -0,0 +1,345 @@
+			break;
+	}
+
+	/* No pending signals were queues - try again */
```
Typo: "queued" I assume

andreittr commented on lib/posix-process/signal/signalfd.c:
```diff
@@ -0,0 +1,345 @@
+
+static struct uk_file *uk_sigfile_create(const uk_sigset_t *mask,
+					 __sz masksz __unused,
+					 int flags __unused)
```
(1) If `masksz` and `flags` are unused, and this is a static function, why keep them?
(2) If this is a static func, why is it prefixed with `uk_` unlike other sigfile funcs?
(3) You're a file driver that's including libc types by design, why are you using `__sz`? Is that mandated by some other code this interacts with?

Did you intend for this func to be public (at some future point)?

Also: If you intend to export the sigfile API, consider adding a
```c
void uk_sigfile_set_mask(const struct uk_file *f, const uk_sigset_t *mask);
```

mogasergiu replied:
for 1) and 2), they are artifacts from the time i was initially writing this

3) I'd rather use our own internal types and use libc types in the syscall entry wrappers and convert if needed. Maybe I should add a cast just to make it clear, even if it does nothing. I want to use our internal types internally because it is inevitable that we will call kernel internal functions.

andreittr replied:
That's fair; if this func stays private and you don't intend to export the sigfile API, feel free to use whatever mix of types suits you best.
I'm only gatekeeping the public `uk_file`-related APIs to use userspace types; everyone else carry on :smile:

mogasergiu replied:
Perfectly justified, agreed.

andreittr commented on lib/posix-process/signal/signalfd.c:
```diff
@@ -0,0 +1,345 @@
+	uk_sigdelset(&priv->sigf.mask, SIGKILL);
+	uk_sigdelset(&priv->sigf.mask, SIGSTOP);
+
+	pproc = uk_pprocess_current();
```
(This is assuming you intend to make this function public; disregard if it stays private)

Might be worth moving this higher up in the function and erroring/asserting that this func is called only in a pprocess context, and not from a raw kernel thread.

mogasergiu replied:
Will stay private.

andreittr commented on lib/posix-process/signal/signalfd.c:
```diff
@@ -0,0 +1,345 @@
+int signalfd(int fd, const sigset_t *mask, int masksz)
+{
+	return (int)uk_syscall_e_signalfd((long)fd, (long)mask, (long)masksz);
+}
```
Hmm, my manpages declare libc `signalfd` as
```c
#include <sys/signalfd.h>

       int signalfd(int fd, const sigset_t *mask, int flags);
```
not `int masksz`, I think you're misusing the args and it should be
```c
return (int)uk_syscall_e_signalfd4((long)fd, (long)mask, (long)(sizeof(*mask)), flags);
```

mogasergiu replied:
No idea what happened, I guess miscopied from the signature of signalfd4.

andreittr replied:
The wrapper is still wrong after the change, because you had the wrong signature for the `signalfd` syscall; see previous comment / code immediately below. My previously suggested version should work correctly though.

andreittr commented on lib/posix-process/signal/signalfd.c:
```diff
@@ -0,0 +1,345 @@
+#if CONFIG_LIBPOSIX_FDTAB
+static int uk_sys_signalfd_create(const uk_sigset_t *mask,
+				  __sz masksz,
+				  int flags)
```
A few words regarding the `uk_sys_*` namespace (I am assuming you're preparing to have these public one day, otherwise pls dump the `uk_sys_` prefix entirely):
It's intended to contain Unikraft-internal versions of actual syscalls that have the exact same semantics, modulo using the appropriate kernel refs instead of user refs (e.g., `const struct uk_file *` or `uk_ofile *` instead of `int fd`) where possible. 
The naming should also match existing syscalls (with some exceptions for silly things and exceptions on top of those), but generally: if there's no syscall, you don't get a `uk_sys_*` name; you have to use your own namespace.

So in this specific instance, there's only the `signalfd` syscall that does everything (technically `signalfd4`, but that falls under the aforementioned "silly things"), so what one would expect here is just
```c
int uk_sys_signalfd(int fd, const uk_sigset_t *mask, size_t masksz, int flags);
```
The two helpers IMO should not be exposed under `uk_sys_`, as they have no corresponding syscall.

andreittr commented on lib/posix-process/signal/signalfd.c:
```diff
@@ -0,0 +1,345 @@
+	uk_sigdelset(mask, SIGKILL);
+	uk_sigdelset(mask, SIGSTOP);
+
+	return 0;
```
I believe you forgot to release the open file after you're done (`uk_fdtab_ret` before #1602 , `uk_ofile_release` after).

andreittr reviewed (commented):
A rough first pass over only the file parts of the PR. There are some broader issues with synchronization and lifetime, but I'll leave those for a more in-depth review after a rebase.

mogasergiu commented:
Force pushed rebase

mogasergiu committed:
lib/ukfile: Add `uk_iov_remaining` to help tell remaining iov space

Add a new routine `uk_iov_remaining` that simply checks how many bytes
the memory regions described by `iov[iovcnt]`, starting at `cur` offset
from the buffer at `iov[iovi]` can fit.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

andreittr commented:
Conflicts/interacts with #1602 , whichever merges first requires patches to the other.

andreittr commented on lib/posix-process/signal/signalfd.c:
```diff
@@ -0,0 +1,342 @@
+UK_LLSYSCALL_R_DEFINE(int, signalfd,
+		      int, fd,
+		      const uk_sigset_t *, mask,
+		      int, flags)
```
The args for the signalfd syscall are:
```c
int ufd, sigset_t *user_mask, size_t sizemask
```
Only `signalfd4` takes flags; both of them take `sizemask` as 3rd arg. [Docs for reference](https://www.man7.org/linux/man-pages/man2/signalfd.2.html#VERSIONS)

andreittr commented on lib/posix-process/signal/signalfd.c:
```diff
@@ -0,0 +1,342 @@
+		      size_t, masksz,
+		      int, flags)
+{
+	return uk_sys_signalfd(fd, mask, flags);
```
It's OK if our internal signal implementation has a fixed size for a sigset and we do not pass `masksz` on; however, because this value is coming from userspace, we should at the very least validate it (`masksz >= sizeof(uk_sigset_t)`) to prevent over-reads.

(As a tangent, `if (unlikely(!mask)) return -EFAULT;` wouldn't hurt either. Might make sense to have `uk_sys_signalfd` take the masksz and perform these validations itself)

andreittr commented on lib/posix-process/signal/signal_file.c:
```diff
@@ -0,0 +1,334 @@
+		      int, flags)
+{
+	if (unlikely(masksz != sizeof(uk_sigset_t)))
+		return -EINVAL;
```
This condition will break in bincompat mode; both glibc and musl define a sigset that's 128 bytes / 1024 bits long (funnily enough both of them as a `struct` containing a `long` array) and will pass that exact 128 byte masksz to your syscall.
That's why I said to validate that `masksz >= sizeof(uk_sigset_t)` not exact equality.

mogasergiu replied:
The thing with masksz, I am not sure how ok this is. There is some incompatibility somewhere between how we do it and what the mask should look like maybe.

andreittr replied:
Oof, disregard, it seems that libcs define larger sigsets than the kernel supports for various reasons.
That would be fine, if they passed the size of that struct onto the kernel; they don't, they always pass an arch-specific constant that the kernel checks _by exact value_.
That would be fine if just an observed behavioral quirk, but unfortunately it's [documented behavior](https://www.man7.org/linux/man-pages/man2/sigwaitinfo.2.html#VERSIONS) for signal syscalls to receive the _exact_ value the kernel expects; even if the documentation is wrong about what exact value that is :roll_eyes: .

andreittr commented on lib/posix-process/signal/signal_file.c:
```diff
@@ -0,0 +1,334 @@
+int signalfd(int fd, const sigset_t *mask, int flags)
+{
+	return (int)uk_syscall_e_signalfd((long)fd, (long)mask, (long)flags);
+}
```
The libc wrapper is still wrong yet again, please fix as advised.

mogasergiu replied:
Oops, forgot to update this one.

andreittr commented on lib/posix-process/signal/signal_file.c:
```diff
@@ -0,0 +1,334 @@
+		      size_t, masksz)
+{
+	if (unlikely(masksz != sizeof(uk_sigset_t)))
+		return -EINVAL;
```
Same issue with `masksz` validation as above. Also, unnecessary code duplication.

andreittr commented on lib/posix-process/signal/signal_file.c:
```diff
@@ -0,0 +1,334 @@
+	int fd;
+
+	if (unlikely(!mask))
+		return -EFAULT;
```
Why do you only check `mask` here and not in `signalfd_set_mask`? Or, as previously advised, in the common wrapper?

mogasergiu commented:
Force-push rebase following merge of #1602 
+ `s/uk_fdtab_ret/uk_ofile_release` in `signal_file.c`

mogasergiu committed:
lib/posix-fdtab: Fix docstring of `uk_fdtab_open`

Commit 5050e9f11e57 ("lib/posix-fd: Move heap-alloc ofiles out of fdtab")
did `s/uk_fdtab_ret/uk_ofile_release` across the codebase but forgot to
update the comment for `uk_fdtab_open`. Do so now, indicating to use
`uk_ofile_release` instead.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

andreittr commented on lib/posix-process/signal/signal_file.c:
```diff
@@ -0,0 +1,354 @@
+	if (fd == -1)
+		return signalfd_create(mask, masksz, flags);
+
+	return signalfd_set_mask(fd, mask, masksz);
```
Is there a reason we don't validate `mask` and `masksz` here in `uk_sys_signalfd` instead of duplicated checks in `_create` and `_set_mask`? AFAICT neither of the two are called anywhere else and likely don't need independent arg validation.

mogasergiu replied:
No reason, just laziness/copy-paste.

andreittr commented on lib/nolibc/musl-imported/include/sys/signalfd.h:
```diff
@@ -0,0 +1,49 @@
+#ifndef _SYS_SIGNALFD_H
```
Could you please add a few lines to the commit message importing this file detailing its provenance? It really helps us keep track of imported code. You can see examples in the commit history, but tl;dr something like:
```
Source of import:
Repository: https://git.musl-libc.org/cgit/musl
Tag: v1.2.4 (f5f55d65)
Path: include/sys/signalfd.h
```
(adapted to whatever origin & tag you used)

andreittr commented on lib/posix-fdtab/include/uk/posix-fdtab.h:
```diff
@@ -69,7 +69,7 @@ int uk_fdtab_open_desc(struct uk_ofile *of, unsigned int mode);
  * Gets the open file description associated with descriptor `fd`.
  *
- * Users should call uk_fdtab_ret when done with the open file reference.
+ * Users should call uk_ofile_release when done with the open file reference.
```
Leave it to API users to find bugs in documentation :smile: cheers for this.

andreittr commented on lib/posix-process/signal/signal.h:
```diff
@@ -295,6 +303,62 @@ void pprocess_signal_arch_get_ucontext(ucontext_t *ucontext,
+#if CONFIG_LIBPOSIX_PROCESS_SIGNAL
+#if CONFIG_LIBPOSIX_PROCESS_SIGNALFD
+static inline
+bool pprocess_signal_sendable(struct posix_process *pproc, int signum)
```
I appreciate you factored this out from its callsites in `signal.c`, although I believe you could have gone even further. Considering that both callsites negate the result, consider pushing that negation into the function itself, turning it into `pprocess_signal_masked`/`_dropped`/`_muted`/whatever better name you come up with.
This will also remove the slightly head-scratching double-negation when signalfds are disabled, and the callsites read as `if (!(!IS_IGNORED))`.

mogasergiu replied:
Let's go with should_drop

andreittr commented on lib/posix-process/signal/signal.h:
```diff
@@ -295,6 +303,62 @@ void pprocess_signal_arch_get_ucontext(ucontext_t *ucontext,
+				 int signum __unused)
+{
+	/* TODO: Set file event for one specific thread only */
+}
```
While we don't yet have thread-specific file event notification, I believe we should until then notifiy all signalfd-waiting threads instead of doing nothing.
If we do nothing, a thread sleeping on a signalfd may never wake up, even as signals are queued to it, possibly breaking apps. The cost of needlessly waking up all other signalfd threads to immediately read an `-EAGAIN` is IMO worth it.
(Technical nit: since signalfds are explicitly polled files, threads may not even enter `read` on spurious wakeups, rather immediately `poll` and block again -- an even shorter path)

andreittr commented on lib/posix-process/signal/signal_file.c:
```diff
@@ -0,0 +1,354 @@
+		if (!uk_sigismember(&sigf->mask, sn) ||
+		    (!IS_PENDING(pthread->signal->sigqueue, sn) &&
+		     !IS_PENDING(pproc->signal->sigqueue, sn)))
+			continue;
```
Might I suggest
```c
pprocess_signal_foreach(sn)
    if (uk_sigismember(&sigf->mask, sn) &&
        (IS_PENDING(pthread->signal->sigqueue, sn) ||
         IS_PENDING(pproc->signal->sigqueue, sn)))
        return UKFD_POLLIN;

return 0;
```
it's the same logic, but IMO much easier to parse without all those negations and control flow jumps.

mogasergiu replied:
This is so not my style but ok :laughing:

andreittr commented on lib/posix-process/signal/signal_file.c:
```diff
@@ -0,0 +1,354 @@
+static void si_to_ssi(siginfo_t *si, struct signalfd_siginfo *ssi)
+{
+	/* Zero out the structure first */
+	memset(ssi, 0, sizeof(*ssi));
```
Do we strictly need this? As in, are there fields of `ssi` that we might not initialize and there is documented (or at least observed) behavior that signalfd syscalls zero out the rest?

It's not a rhetorical question (I've been burned by this in `stat` code). If we _need_ this zeroing out we should document it, so some eager future dev doesn't optimize it away. If we don't need it, we shouldn't bother.

mogasergiu replied:
Yep, I did some tests on linux userspace and it seems the unused members are always zeroed out.

andreittr commented on lib/posix-process/signal/signal_file.c:
```diff
@@ -0,0 +1,354 @@
+
+		si_to_ssi(&sig->siginfo, &si);
+		uk_signal_free(pthread->_a, sig);
+		iovlen_wr += uk_iov_scatter(iov, iovcnt,
```
Oof, you're doing double the memwrites here (using `si` as tmp), and I don't think the compiler is smart enough to optimize it away.
I do get _why_ you're doing it like this, and don't see a clean way of doing both type conversion and iovec scatter any other way (yet).
I'll make a mental note for `iovutil.h` :thinking:

andreittr commented on lib/posix-process/signal/signal_file.h:
```diff
@@ -0,0 +1,164 @@
+ * @return
+ *   Pointer to the created signal file or NULL on creation failure
+ */
+struct uk_file *signal_file_create(struct uk_alloc *a,
```
I recommend you return a `const struct uk_file *` to enforce the rule that all ukfiles, once created, are immutable.
I know this is a lib-internal header/API, but I don't want any precedent out there for mutable ukfile refs being passed around after being made public.

andreittr commented on lib/posix-process/signal/signal_file.c:
```diff
@@ -0,0 +1,354 @@
+	 */
+	uk_sigcopyset(&sigf->mask, new_mask);
+	uk_sigdelset(&sigf->mask, SIGKILL);
+	uk_sigdelset(&sigf->mask, SIGSTOP);
```
I believe you forgot to update the pproc allmask when an existing signalfd's mask is changed (you need to at least OR in any newly set bits to prevent missed signal deliveries; cleared bits can be updated by the usual pruning runs)

mogasergiu replied:
oh good catch!

andreittr commented on lib/posix-process/signal/signal_file.h:
```diff
@@ -0,0 +1,164 @@
+{
+	struct uk_signal_file *it, *itn;
+
+	uk_sigemptyset(&ctx->allmask);
```
Let's hope the allmask is never read while this function is executing, else we'll miss signals; especially since `uk_file_event_set` can end up doing callbacks.
I understand synchronization is not in scope but this still feels really scary; are there huge performance drawbacks to using a local temp to gather the new allmask before writing it in before returning?

mogasergiu replied:
None, I even did it like that while updating but then I decided to leave it for when we actually do synchronization. But now that makes two of us that think it might be a good idea so I'll do it :small_airplane:

andreittr reviewed (changes requested):
Another pass, with a closer-toothed comb; it's shaping up nicely in general.

Also, I assume synchronization is not in scope at all and will be addressed in future work.

andreittr commented on lib/posix-process/signal/signal_file.h:
```diff
@@ -0,0 +1,167 @@
+ *   Signal mask to be monitored by the signal file
+ *
+ * @return
+ *   Pointer to the created signal file or NULL on creation failure
```
This line contradicts implementation where you return errptrs.
Judging by what this function does though, failing to allocate memory is really the only way it _can_ fail, so returning NULL may well be valid :shrug: 
I don't have strong feelings either way for NULL or ERRPTR, but please have the docs and impl be in line.

andreittr commented on lib/posix-process/signal/signal_file.h:
```diff
@@ -0,0 +1,167 @@
+		uk_file_event_set(&it->f, UKFD_POLLIN);
+	}
+
+	uk_sigorset(&ctx->allmask, &allmask);
```
I think this should be `uk_sigcopyset`, as ORing `allmask` in will not clear any pruned signals.

andreittr commented on lib/posix-process/signal/signal_file.h:
```diff
@@ -0,0 +1,167 @@
+ *   Containing uk_signal_file structure
+ */
+static inline
+struct uk_signal_file *uk_file_to_signal_file(const struct uk_file *f)
```
I believe you can move this over into `signal_file.c`, as no other inline in this header is using it AFAICT.
And if you move it, you can add a `UK_ASSERT(FILE_IS_SIGNAL(f))` for the extra safety against using garbage data. As a bonus, you get consistent asserts for all ops (poll, read, set_mask, and destructor) that touch the `uk_signal_file` struct without explicit checks.

andreittr reviewed (commented):
Pass 3, getting pretty close to a checkmark from my side.

andreittr commented on lib/posix-process/signal/signal_file.c:
```diff
@@ -0,0 +1,358 @@
+}
+
+/* Convert internal siginfo_t to signalfd_siginfo structure */
+static void si_to_ssi(siginfo_t *si, struct signalfd_siginfo *ssi)
```
`const siginfo_t *si` perhaps? Won't change code generation, but might save us debug cycles if someone ever mistypes `si` instead of `ssi` in a future patch :smile:

andreittr commented on lib/posix-process/signal/signal_file.h:
```diff
@@ -0,0 +1,151 @@
+
+	uk_sigemptyset(&allmask);
+
+	uk_list_for_each_entry_safe(it, itn, &ctx->sigfiles, sigf_head) {
```
I _think_ you can use the plain `uk_list_for_each_entry` here; IIRC the safe versions are for when you intend to modify the list mid-traversal. Saves the `itn` tempvar.

andreittr commented on lib/posix-process/signal/signal_file.c:
```diff
@@ -0,0 +1,358 @@
+	struct posix_thread *pthread;
+	struct posix_process *pproc;
+	struct uk_signal_file *sigf;
+	struct signalfd_siginfo si;
```
Maybe rename `si` to `ssi` for better name consistency with `si_to_ssi()`?

andreittr reviewed (commented):
Ok, (hopefully) final pass, only trivial things left.

andreittr reviewed (approved):
Cheers! Looks good; can't wait to see the synchronization PR :smile: 

Reviewed-by: Andrei Tatar <andrei@unikraft.io>

michpappas commented:
Reviewed-by: Michalis Pappas <michalis@unikraft.io>
Approved-by: Michalis Pappas <michalis@unikraft.io>
